fairscale/experimental/nn/data_parallel/gossip/distributed.py

```diff
@@ -0,0 +1,1192 @@
+                    p_fp16.copy_(p_fp32)
+        ef_copy_rec.stop()
+        self.logger.debug("Error feedback copy-back completed")
+
+    def _maybe_sgp_overlap_pre_communicate_error_feedback(
+        self, fp16_fp32_list: List[Tuple[torch.Tensor, torch.Tensor]]
+    ) -> None:
+        if self._should_perform_sgp_overlap() and fp16_fp32_list:
+            # Initialize error feedback for SGP-overlap
+            if self.ef1 is None:
+                self.ef1 = [p_fp32.clone().detach_() for _, p_fp32 in fp16_fp32_list]
+
+            with torch.no_grad():
+                assert self.ef1 is not None
+                for ef1, (p_fp16, p_fp32) in zip(self.ef1, fp16_fp32_list):
+                    ef1.copy_(p_fp32 - p_fp16.float())
+
+    def perform_slowmo(self, optimizer: torch.optim.Optimizer, fp32_params: Optional[torch.Tensor] = None) -> None:
+        """ This is to be called after optimizer.step(). It performs the approximate averaging using
+        the base algorithm (SGP/ LocalSGD) and the slow momentum step. Since LocalSGD and the slow
+        momentum step are not performed every iteration, it only performs those when needed.
+
+        Args:
+            optimizer (torch.optim.Optimizer): The optimizer being used for training the model
+            fp32_params (Optional[torch.Tensor]: To be used when performing fp16 training. Needs to be
+                        set to the fp16 copy of the parameters (default: None)
+        """
+        # Done here in case the global momentum buffers have not been initialized by the caller.
+        # In an ideal implementation, this would be called by the caller. We do it here instead of
+        # waiting for it to happen in the global_momentum step function so that we store a copy of
+        # the version of the parameters at iteration 0 and can use them for a slow momentum step later.
+        if not self.global_momentum_buffers_initialized:
+            self._init_global_momentum_buffers(optimizer)
+
+        fp16_fp32_list = list(self._fp16_fp32_iterator(optimizer, fp32_params))
+        self.logger.debug("Created a list of fp16 and fp32 corresponding parameters")
+
+        self.logger.debug(
+            "Booleans set. Values - self._should_perform_slowmo()=%r, self._should_perform_localsgd()=%r, self._should_allreduce_params()=%r",
+            self._should_perform_slowmo(),
+            self._should_perform_localsgd(),
+            self._should_allreduce_params(),
+        )
+        self.logger.debug("Step number(0-indexed)=%d", self.num_updates)
+
+        if (
+            self.num_updates == 0
+            and fp32_params is None
+            and not hasattr(optimizer, "_amp_stash")
+            and any(p.dtype == torch.float16 for p in self.parameters())
+        ):
+            self.logger.warning("WARNING: please set fp32_params in perform_slowmo() in order to avoid accuracy loss")
+
+        self._maybe_pre_communicate_error_feedback(fp16_fp32_list)
```

Comment:
not an issue, but just linking to another way of doing this I think: there's effectively a state machine here, FSDP makes it explicit ([see](https://github.com/facebookresearch/fairscale/blob/main/fairscale/nn/data_parallel/fully_sharded_data_parallel.py#L65)), maybe easier to debug at times (not really needed now that it works) or to save/restore a state